Adds a new command to randomize (shuffle) tracks

jarv opened the pull request:
Command added to the library, playlist and queue view for
feature request https://github.com/cmus/cmus/issues/231.

flyingmutant commented:
Thanks for the PR! I'll add some comments inline.

flyingmutant commented on command_mode.c:
```diff
@@ -1352,6 +1352,26 @@ static void cmd_pwd(char *arg)
+	editable_unlock();
+}
+
+
```
Please keep whitespace between functions consistent with status quo in files you change, which is 1 empty line, not 2.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
 	list_del(&track->node);
 }
 
+static void shuffle_array(void *array, size_t n, size_t size) 
```
I think it is better to move this function to `misc.c`.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
 
+static void shuffle_array(void *array, size_t n, size_t size) 
+{
+	char tmp[size];
```
Hm, is this the first place in cmus where var-sized arrays are used? I am not 100% sure I am comfortable with it for some reason.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
+{
+	char tmp[size];
+	char *arr = array;
+	size_t stride = size * sizeof(char);
```
`sizeof(char)` is 1.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
+	size_t stride = size * sizeof(char);
+	if (n > 1) {
+		size_t i;
+		for (i = 0; i < n - 1; ++i) {
```
These 3 lines can be replaced by something like `for (ssize_t i = 0; i < (ssize_t)n - 1; ++i)`.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
+{
+	struct list_head *item, *tmp;
+	struct rb_root tmp_tree = RB_ROOT;
+	struct simple_track **track_list;
```
`track_list` is really an array, not a list.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
+	list_for_each(item, head) {
+		len++;
+	}
+	track_list = xmalloc(len * sizeof(struct simple_track *));
```
`sizeof(track_list[0])` is much better.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
+		track_list[track_cnt] = track;
+		track_cnt++;
+	}
+	shuffle_array(track_list, len, sizeof(struct simple_track *));
```
Same here – no need to hardcode exact types in `sizeof`.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
+		track_cnt++;
+	}
+	shuffle_array(track_list, len, sizeof(struct simple_track *));
+	for (i=0; i<len; i++) {
```
No need to declare `i` beforehand – just do `for (int i = 0; …)`.

flyingmutant commented on track.c:
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
+
+	tree_root->rb_node = tmp_tree.rb_node;
+	_list_add(head, tmp_head.prev, tmp_head.next);
+
```
Extra newline.

jarv commented:
I've gone ahead and addressed your comments in a new commit (let me know if you would like me to squash).  For the var-sized array I was hesitant to allocate MAX_TRACKS and avoided doing something more complex for the shuffle. Let me know what you think if you are still open to taking this. I use CMUs all the time and randomizing tracks has been very useful for me when DJing from a playlist.

jarv commented:
Updated to fix the last comments

flyingmutant commented:
Rebased and merged in a10327e7dbb07fcc68e3fd08ff519f8bffb2353b, thanks a lot!
